### PR TITLE
Re-implement Workspace.isResolutionRequired to use the new dependency resolver

### PR DESCRIPTION
**Pull request by hartbit**

### CODE REVIEW

File: Sources/SPMTestSupport/TestWorkspace.swift [hunk 1]
```diff
@@ -38,7 +38,7 @@ public final class TestWorkspace {
         packages: [TestPackage],
         toolsVersion: ToolsVersion = ToolsVersion.currentToolsVersion,
         skipUpdate: Bool = false,
-        enablePubGrub: Bool = false
+        enablePubGrub: Bool = true
```
Comment (aciidb0mb3r) [inline Sources/SPMTestSupport/TestWorkspace.swift]: Cool! I had a task for doing this, heh. Can you make this change in a separate PR?

File: Sources/SPMTestSupport/TestWorkspace.swift [hunk 2]
```diff
@@ -288,6 +296,42 @@ public final class TestWorkspace {
         result(graph, diagnostics)
     }
 
+    public func checkIsResolutionRequired(
+        pins: [PackageReference: CheckoutState],
+        managedDependencies: [ManagedDependency],
+        _ check: (Bool, DiagnosticsEngine) -> ()
```
Comment (aciidb0mb3r) [inline Sources/SPMTestSupport/TestWorkspace.swift]: maybe add labels to the types for clarity ?

File: Sources/Workspace/LocalContainerProvider.swift [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+import SourceControl
+
+/// PackageContainerProvider implementation used by Workspace to do a dependency pre-calculation using the cached
+/// dependency information (Workspace.DependencyManifests) to check if dependency resolution is required before
+/// performing a full resolution.
+struct LocalContainerProvider: PackageContainerProvider {
+    let root: PackageGraphRoot
+    let currentManifests: Workspace.DependencyManifests
+    let config: SwiftPMConfig
+    let diagnostics: DiagnosticsEngine
```
Comment (aciidb0mb3r) [inline Sources/Workspace/LocalContainerProvider.swift]: doc comments pls

File: Sources/Workspace/LocalContainerProvider.swift [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    func getContainer(
+        for identifier: PackageReference,
+        skipUpdate: Bool,
+        completion: @escaping (Result<PackageContainer, AnyError>) -> Void
+    ) {
+        if let manifest = currentManifests.dependencies.first(where: { $1.packageRef == identifier }) {
```
Comment (aciidb0mb3r) [inline Sources/Workspace/LocalContainerProvider.swift]: Maybe add comment explaining what you're doing here?

File: Sources/Workspace/LocalContainerProvider.swift [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            }
+        case .edited, .local:
+            break
+        }
+
+        return []
+    }
+
+    func isToolsVersionCompatible(at version: Version) -> Bool {
+        return true
```
Comment (aciidb0mb3r) [inline Sources/Workspace/LocalContainerProvider.swift]: should we add an assert here?

File: Sources/Workspace/LocalContainerProvider.swift [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    func getDependencies(at version: Version) throws -> [PackageContainerConstraint] {
+        return manifest.dependencyConstraints(config: config)
+    }
+
+    func getDependencies(at revision: String) throws -> [PackageContainerConstraint] {
+        return manifest.dependencyConstraints(config: config)
+    }
+
+    func getUnversionedDependencies() throws -> [PackageContainerConstraint] {
+        return manifest.dependencyConstraints(config: config)
```
Comment (aciidb0mb3r) [inline Sources/Workspace/LocalContainerProvider.swift]: we should add an assert that dependency is nil or edited

File: Sources/Workspace/LocalContainerProvider.swift [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    func isToolsVersionCompatible(at version: Version) -> Bool {
+        return true
+    }
+
+    func versions(filter isIncluded: (Version) -> Bool) -> AnySequence<Version> {
+        return AnySequence(reversedVersions)
+    }
+
+    func getDependencies(at version: Version) throws -> [PackageContainerConstraint] {
+        return manifest.dependencyConstraints(config: config)
```
Comment (aciidb0mb3r) [inline Sources/Workspace/LocalContainerProvider.swift]: we should add an assert that dependency is at this version

File: Sources/Workspace/Workspace.swift [hunk 8]
```diff
@@ -1231,15 +1232,13 @@ extension Workspace {
 
         let currentManifests = loadDependencyManifests(root: graphRoot, diagnostics: diagnostics)
 
-        // Check if a new resolution is required.
-        let dependencies =
-            graphRoot.constraints(config: config) +
-            // Include constraints from the manifests in the graph root.
-            graphRoot.manifests.flatMap({ $0.dependencyConstraints(config: config) }) +
-            currentManifests.dependencyConstraints()
+        let isRequired = isResolutionRequired(
```
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: nit: isRequired -> isResolutionRequired

File: Sources/Workspace/Workspace.swift [hunk 9]
```diff
@@ -1422,84 +1413,80 @@ extension Workspace {
...
-            $0.requirement == constraintSet[$0.identifier]
-        })
+        guard !diagnostics.hasErrors else { return true }
+        guard case .success(let bindings) = result else { return true }
 
         // Otherwise, check checkouts and pins.
-        for constraint in constraintSet {
-            let url = constraint.key.path
-            let dependency = managedDependencies[forURL: url]
+        for (container, boundVersion) in bindings {
```
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: I think we can kill this loop now since we've already checked everything that was needed using the resolver.

File: Sources/SPMTestSupport/TestWorkspace.swift (hunk 1 above)
Comment (hartbit) [reply Sources/SPMTestSupport/TestWorkspace.swift]: Done: https://github.com/apple/swift-package-manager/pull/2421

File: Sources/SPMTestSupport/TestWorkspace.swift (hunk 2 above)
Comment (hartbit) [reply Sources/SPMTestSupport/TestWorkspace.swift]: Done.

File: Sources/Workspace/LocalContainerProvider.swift (hunk 4 above)
Comment (hartbit) [reply Sources/Workspace/LocalContainerProvider.swift]: Done.

File: Sources/Workspace/LocalContainerProvider.swift (hunk 3 above)
Comment (hartbit) [reply Sources/Workspace/LocalContainerProvider.swift]: Done.

File: Sources/Workspace/LocalContainerProvider.swift (hunk 5 above)
Comment (hartbit) [reply Sources/Workspace/LocalContainerProvider.swift]: I did a check against the current tools version.

File: Sources/Workspace/Workspace.swift (hunk 8 above)
Comment (hartbit) [reply Sources/Workspace/Workspace.swift]: Done.

File: Sources/Workspace/Workspace.swift (hunk 9 above)
Comment (hartbit) [reply Sources/Workspace/Workspace.swift]: Actually, we still need this. The `LocalPackageContainer`-backed resolver can still return a successful resolution with different bound versions: for example, when changing a dependency from a version to a branch or revision requirement, the resolver will always prefer the branch or revision requirement and will resolve successfully, without trying to checkout the branch or revision.

Comment (hartbit) [reply Sources/Workspace/Workspace.swift]: Although I admit that I did find it weird that the resolver would resolve a dependency to a branch without checking it out to see if it had different dependencies.

Comment (hartbit) [reply Sources/Workspace/Workspace.swift]: Btw, if you are curious, you can see this in action by commenting out the loop and you will see tests `WorkspaceTests.testRevisionVersionSwitch` and `WorkspaceTests.testLocalVersionSwitch` fail. But those tests pass dependencies as package root dependencies, not as dependencies of the root manifest.

File: Sources/Workspace/LocalContainerProvider.swift (hunk 6 above)
Comment (hartbit) [reply Sources/Workspace/LocalContainerProvider.swift]: Done.

File: Sources/Workspace/LocalContainerProvider.swift (hunk 7 above)
Comment (hartbit) [reply Sources/Workspace/LocalContainerProvider.swift]: Done.

File: Sources/Workspace/LocalContainerProvider.swift (hunk 6 above)
Comment (hartbit) [reply Sources/Workspace/LocalContainerProvider.swift]: Actually, I don't think we want. If a dependency requirement changes from version/revision to edited/local, the provider will still return it, and this function will be called. That's what the loop after resolution to diff the bound version with dependency and pins is useful. What do you think?

File: Sources/Workspace/LocalContainerProvider.swift (hunk 7 above)
Comment (hartbit) [reply Sources/Workspace/LocalContainerProvider.swift]: Same comment as above, but the other way round.

**Comment by hartbit** [comment]
The following PR https://github.com/apple/swift-package-manager/pull/2421 needs to be merged first.

File: Sources/Workspace/LocalContainerProvider.swift [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+import PackageModel
+import PackageGraph
+import TSCBasic
+import TSCUtility
+import SourceControl
+
+/// PackageContainerProvider implementation used by Workspace to do a dependency pre-calculation using the cached
+/// dependency information (Workspace.DependencyManifests) to check if dependency resolution is required before
+/// performing a full resolution.
+struct LocalContainerProvider: PackageContainerProvider {
```
Comment (aciidb0mb3r) [inline Sources/Workspace/LocalContainerProvider.swift]: Maybe call it `ResolverPrecomputationProvider`?

File: Sources/Workspace/LocalContainerProvider.swift [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+                dependency: manifest.dependency,
+                config: config,
+                currentToolsVersion: currentToolsVersion
+            )
+
+            return completion(.success(container))
+        }
+
+        // Continue searching from the Workspace's root manifests.
+        if let index = dependencyManifests.root.packageRefs.firstIndex(of: identifier) {
```
Comment (aciidb0mb3r) [inline Sources/Workspace/LocalContainerProvider.swift]: Maybe add a FIXME saying that we should use a dictionary for faster lookups

File: Sources/Workspace/LocalContainerProvider.swift [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+                dependency: nil,
+                config: config,
+                currentToolsVersion: currentToolsVersion
+            )
+
+            return completion(.success(container))
+        }
+
+        // As we don't have anything else locally, error out.
+        completion(.init(Diagnostics.fatalError))
```
Comment (aciidb0mb3r) [inline Sources/Workspace/LocalContainerProvider.swift]: does this ever happen?

File: Sources/Workspace/LocalContainerProvider.swift [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+    }
+}
+
+private struct LocalPackageContainer: PackageContainer {
+    let manifest: Manifest
+    let dependency: ManagedDependency?
```
Comment (aciidb0mb3r) [inline Sources/Workspace/LocalContainerProvider.swift]: Maybe add  doc comments here explaining that this property is only present for remote packages.

File: Sources/Workspace/LocalContainerProvider.swift [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    let config: SwiftPMConfig
+    let currentToolsVersion: ToolsVersion
+
+    var identifier: PackageReference {
+        if let identifier = dependency?.packageRef {
```
Comment (aciidb0mb3r) [inline Sources/Workspace/LocalContainerProvider.swift]: Maybe add a comment like:

```
// Get package reference from the managed dependency or compute it for local packages.
```

File: Sources/Workspace/LocalContainerProvider.swift (hunk 12 above)
Comment (hartbit) [reply Sources/Workspace/LocalContainerProvider.swift]: I think I remember this happens when you add a new dependency as it doesn't exist locally.

File: Sources/Workspace/Workspace.swift (hunk 9 above)
Comment (aciidb0mb3r) [reply Sources/Workspace/Workspace.swift]: Hmm, that seems wrong. We should throw from `getUnversionedDependencies` if it's accessed for non-versioned dependencies.

File: Sources/Workspace/LocalContainerProvider.swift (hunk 6 above)
Comment (aciidb0mb3r) [reply Sources/Workspace/LocalContainerProvider.swift]: Ah. I think we need to throw from here then we would be able to kill the loop: https://github.com/apple/swift-package-manager/pull/2416/files#r349707709

File: Sources/Workspace/LocalContainerProvider.swift (hunk 12 above)
Comment (aciidb0mb3r) [reply Sources/Workspace/LocalContainerProvider.swift]: ah.

File: Sources/Workspace/LocalContainerProvider.swift (hunk 13 above)
Comment (hartbit) [reply Sources/Workspace/LocalContainerProvider.swift]: I'll add a comment, but its also present for local packages. Its only missing for root packages.

File: Sources/Workspace/LocalContainerProvider.swift (hunk 14 above)
Comment (hartbit) [reply Sources/Workspace/LocalContainerProvider.swift]: I added a comment, but tweaked it to clarify that its only computed for root packages.

File: Sources/Workspace/LocalContainerProvider.swift (hunk 6 above)
Comment (hartbit) [reply Sources/Workspace/LocalContainerProvider.swift]: You are correct.

File: Sources/Workspace/Workspace.swift (hunk 9 above)
Comment (hartbit) [reply Sources/Workspace/Workspace.swift]: That's correct.

**Comment by hartbit** [comment]
@aciidb0mb3r I resolved all comments. Indeed, despite what I initially thought, throwing from the getDependencies functions is enough to completely remove the need for the loop I had.

**Comment by hartbit** [comment]
@swift-ci please smoke test
